util/rateLimit.go

```diff
@@ -0,0 +1,501 @@
+// Copyright (C) 2019-2022 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package util
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/algorand/go-algorand/util/metrics"
+	"github.com/algorand/go-deadlock"
+)
+
+// ElasticRateLimiter holds and distributes capacity through capacityQueues
+// Capacity consumers are given an error if there is no capacity available for them,
+// and a "capacityGuard" structure they can use to return the capacity when finished
+type ElasticRateLimiter struct {
+	MaxCapacity            int
+	CapacityPerReservation int
+	sharedCapacity         capacityQueue
+	capacityByClient       map[ErlClient]capacityQueue
+	clientLock             deadlock.RWMutex
+	// CongestionManager and enable flag
+	cm                       CongestionManager
+	enableCM                 bool
+	congestionControlCounter *metrics.Counter
+}
+
+// ErlClient clients must support OnClose for reservation closing
+type ErlClient interface {
+	OnClose(func())
+}
+
+// capacity is an empty structure used for loading and draining queues
+type capacity struct {
+}
+
+// Capacity Queue wraps and maintains a channel of opaque capacity structs
+type capacityQueue chan capacity
+
+// ErlCapacityGuard is the structure returned to clients so they can release the capacity when needed
+// they also inform the congestion manager of events
+type ErlCapacityGuard struct {
+	client ErlClient
+	cq     capacityQueue
+	cm     *CongestionManager
+}
+
+// Release will put capacity back into the queue attached to this capacity guard
+func (cg *ErlCapacityGuard) Release() error {
+	if cg.cq == nil {
+		return nil
+	}
+	select {
+	case cg.cq <- capacity{}:
+		return nil
+	default:
+		return fmt.Errorf("could not replace capacity to channel: %v", cg.cq)
```

Comment:
please add a new err type here and below

Reply:
what is a point printing a chan of empty structs?
when I do I get
```
	a := make(chan struct{}, 10)
	a <- struct{}{}
	a <- struct{}{}
	fmt.Printf("%v\n", a)

0xc00006e060
```
does not look useful at all

Reply:
I agree, will remove.